World multithreading

Janmm14 opened the pull request:
Closes #715 
Closes #728 
Closes #751 

Documentation PR: https://github.com/gomint/gomint-docs/pull/28

I'd like to get some hints on what else needs to be made multithread-ready.

### TODO:

- [x] World events (join, leave, load, unload)
- [x] Command API
- [x] Console commands
- [x] permission ticking
- [x] Easy, standardized API for plugins to read world allow-/blocklist out of config
- [x] Scoreboards ?!
- [x] world-limited event listeners
- [x] External documentation

### Current design changes to #715:
- Only one network manager, but it only ticks worldless connections, otherwise the world ticks the connections
- New connections are first handled on the old main thread until player joined a world
- Chat stays on each world's thread, I think that's better than an extra chat thread.

Also contains a bunch of various fixes as well as some javadoc and code layout fixes.

Janmm14 committed:
World switching now does schedule the target world part in target world thread

Janmm14 committed:
more multithread warnings, PluginWorldSchedulerAdapter stores only a WeakReference, clear syncTaskManager on world unload

Janmm14 committed:
Move to controlled network manager handoff for player connections. Improve concurrency in NetworkManager.

Janmm14 committed:
refactoring gomint from one main thread to one thread per world

- split sync and async scheduler api and implementation
- add per-world sync scheduler
- move ticking code from GoMintServer to each world
- create thread in world constructor
- add plugin instance field to plugin class loader and getCallerPluginLoader() method in CallerDetector for world#scheduler() method returning plugin-based
- replace main thread check with world thread check
- synchronize access to Entity#world
- schedule connection close handling onto world thread

Janmm14 committed:
refactoring gomint from one main thread to one thread per world, vol. 2

- remove sync scheduling from PacketLoginHandler as disconnect is thread safe now
- remove constructor arguments from PacketHandlers
- update world of player to spawnLocation world of PlayerLoginEvent
- make world manager thread safe, start world thread

Janmm14 committed:
EntityPlayer imports

Janmm14 committed:
Force waiting in WorldManager#close until all worlds are stopped.

Janmm14 committed:
Remove static WorldAdapter logger

Janmm14 committed:
Stop asyncWorkerTask on world close

Janmm14 committed:
Don't complain about async world worker accessing world data async

Janmm14 committed:
Fix world threadId init

Janmm14 committed:
Fix tests

Janmm14 committed:
Outsource event caller class creator code to static helper method

Janmm14 committed:
module-info: Export io.gomint.event.interfaces

Janmm14 committed:
Add world-whitelisted event listeners

geNAZt reviewed (changes requested):
This system should be made with more generic prediction usage. So you can have `registerLisetener(T listener, Predicate<Event> shouldFirePredicate)` and there is no need to store the IntSet (which is a candidate for hash collisions) since you can use the Collection (which should be moved to the Set interface) in SimplePluginManager `eventManager.registerListener(EventListener, Collection<String>)` to 

```
eventManager.registerListener(EventListener, Predicate<Event>)

eventManager.registerListener(EventListener, new Predicate<Event>() {
  boolean test(Event event) {
    return !(event instanceof WorldEvent && worlds != null && !worlds.contains(((WorldEvent) event).world().folder()));
  }
})
```

Janmm14 committed:
Remove server tps and server currentTickTime and switch to world's currentTickTime where used

Janmm14 committed:
EventManager now takes a predicate to check if method should be invoked. PluginManager now takes a ConcurrentHashMap.KeySetView as worlds whitelist to allow (concurrent) editing of the whitelist. No defensive copy made anymore.

Janmm14 committed:
Add debug logging about predicate event ignoring

Janmm14 committed:
Javadoc notes that edits to the world set are honored.

Janmm14 committed:
Add WorldLoadEvent and WorldUnloadEvent. Add API for a worlds.yml file in plugin's datafolder for defining active worlds of a plugin, either as whitelist or blacklist.

Janmm14 committed:
Add PlayerWorldJoinEvent and PlayerWorldLeaveEvent.

geNAZt commented on gomint-api/src/main/java/io/gomint/plugin/Plugin.java:
```diff
@@ -206,4 +215,31 @@ public GoMint server() {
+     * @see #activeInWorld(World)
+     */
+    @Nullable
+    public ConcurrentHashMap.KeySetView<String, Boolean> activeWorlds() {
```
What is the general idea behind this? Why does a plugin have active worlds? How is this handled? Should plugins be restricted to those worlds (block setting, world getting, entity manipulations etc.) and unloaded/loaded for each of that worlds?

Janmm14 replied:
The idea is that a plugin *can* and *should* use this to restrict itself to certain worlds. The idea is to use this Set to register its listeners easily to restrict those to certain worlds. See also the automatically updating in the comment at registerListener.

The idea is not to restrict plugins. The idea is that all plugins use this to allow restricting them to certain worlds. By giving plugin authors easy ways for this, as well as providing a standardized way for it (worlds.yml in each plugin's folder), it is likely to be used by plugins.

The idea is not that plugins are loaded per world. 

Where you needed multiple servers connected with bungee in mc java to run not custom made plugins only in parts of your server, the idea is that gomint plugins are all built with this in mind to only run one server with multiple worlds for your minigmae network.

I hope you understand this, I feel like I can't express it correctly.

geNAZt replied:
I see, then maybe this PR should be coupled with a documentation document in https://github.com/gomint/gomint-docs stating that system and its goals.

Janmm14 replied:
Will do

geNAZt commented on gomint-api/src/main/java/io/gomint/plugin/PluginManager.java:
```diff
@@ -72,12 +75,33 @@
+     * @see #registerListener(Plugin, EventListener)
+     * @since 2.0
+     */
+    PluginManager registerListener(Plugin plugin, EventListener listener, @Nullable ConcurrentHashMap.KeySetView<String, Boolean> worlds);
```
I don't like that a specific view is part of the api. Also this should be a Predicate system like i commented earlier in this review

Janmm14 replied:
I thought you only wanted the internal EventManager to have the predicate system.

I required that specific Set implementation to prevent cme's.

geNAZt replied:
No, the API should have that Predicate. The set should not CME, and if it does i would use a immutable set over a specific implementation.

Janmm14 replied:
added register method with predicate and renamed old one with the worlds predicate

geNAZt commented on gomint-api/src/main/java/io/gomint/scheduler/AsyncScheduler.java:
```diff
@@ -0,0 +1,38 @@
+     * @param runnable which should be executed
+     * @return the created and scheduled Task
+     */
+    Task executeAsync(Runnable runnable);
```
Why are methods suffixed with Async? This alreay is AsyncScheduler, can we remove that suffix?

Janmm14 replied:
The normal Scheduler extends this one and adds its sync methods with world parameter. We can also split sync and async schedulers in the API if you prefer that.

geNAZt commented on gomint-server/src/main/java/io/gomint/server/entity/Entity.java:
```diff
@@ -267,11 +268,15 @@ public EntityType type() {
 
     public void world(WorldAdapter world ) {
-        this.world = world;
+        synchronized (this.worldSyncObject) {
```
This should be removed. Setting a world should be forced to only work on the thread of the world

Janmm14 replied:
On the thread of the old world or new world?

geNAZt replied:
the new one

Janmm14 replied:
changed. i added a volatile keyword on he world field though, I think that is needed at least. But I am no master in concurrency

geNAZt replied:
a entity isn't shared between multiple threads is it? so it doesn't need any protections from my POV

Also this should be enforced with the Precondition checks i commented earlier

geNAZt commented on gomint-server/src/main/java/io/gomint/server/entity/EntityPlayer.java:
```diff
@@ -403,41 +453,58 @@ public EntityPlayer teleport(Location to, EntityTeleportEvent.Cause cause) {
+
         // Check if we need to change worlds
         if (!to.world().equals(from.world())) {
+            if (this.world != null) {
```
this should be from.world() and there is no need for a null check

Janmm14 replied:
null check removed

geNAZt commented on gomint-server/src/main/java/io/gomint/server/entity/EntityPlayer.java:
```diff
@@ -403,41 +453,58 @@ public EntityPlayer teleport(Location to, EntityTeleportEvent.Cause cause) {
 
             // Change worlds
             world().removePlayer(this);
             this.world((WorldAdapter) to.world());
```
This needs to be done in to.world().syncScheduler().execute()

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/network/handler/PacketLoginHandler.java:
```diff
@@ -147,171 +139,170 @@ public void handle(PacketLogin packet, long currentTimeMillis, PlayerConnection
-                }
+            // Invalid skin
+            if (!validSkin) {
+                connection.disconnect("Skin is invalid or corrupted");
```
disconnect doesn't work since there is a packet sending safeguard but no world is set. This would NPE in packet sending

Janmm14 replied:
I don't see any world safeguard in packet sending

geNAZt replied:
https://github.com/gomint/gomint/pull/727/files#diff-59e6fa0629aae961d769bd023789e9fe6686ceb54bda8fec9f4b3f216c33cddbR257 This would fire warnings and NPEs (due to the world being null) when dropping a user connection in the packet login handler

Janmm14 replied:
I guess we can just also check world for null then there

geNAZt replied:
Are packet handlers invoked from the world thread where the player is in? Otherwise stuff like settings request will also generate that warning. If they do a `!= null` check seems good

Janmm14 replied:
resolved? I allow packet queueing if world is null now.

geNAZt commented on gomint-server/src/main/java/io/gomint/server/network/handler/PacketLoginHandler.java:
```diff
@@ -147,171 +139,170 @@ public void handle(PacketLogin packet, long currentTimeMillis, PlayerConnection
-                        connection.server().watchdog().done();
-                    });
+            // Update player world
+            player.world((WorldAdapter) player.spawnLocation().world());
```
This should be synced up with the world

Janmm14 replied:
waiting for https://github.com/gomint/gomint/pull/727#discussion_r571447875

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java:
```diff
@@ -149,8 +173,12 @@ protected WorldAdapter(GoMintServer server, File worldDir, String name) {
+        this.syncSchedulerAdapter = new SyncSchedulerAdapter(this.syncTaskManager);
+        this.startAsyncWorker(server.asyncScheduler());
         this.initGamerules();
+        this.thread = new Thread(this::tickLoop, "GoMint World Thread " + worldDir.getName());
```
This thread should come out of a threadpool so when closing the world this thread can be assigned to another world

Janmm14 replied:
I thought thread pools are only relevant for short-living threads.

geNAZt replied:
Thread pools are relevant for all sorts of things, mostly to ensure that we can keep proper track of it on shutdown. There is a thread reaper system which seperates the display of system relevant thread by its thread group: https://github.com/gomint/gomint/blob/8523fd0a0be896eb8fbb4a802b3059ddb5e7498d/gomint-server/src/main/java/io/gomint/server/GoMintServer.java#L626 there should be a group named "gomint-worlds" from which all threads come so we can warn properly

Janmm14 replied:
Seems to me like the only place thread groups are used in whole gomint is at that point. Seems to me like thread grouping (not pooling) is not a thing yet in gomint.

geNAZt replied:
it will be in the future (to ensure that async tasks attached to a plugin get terminated when the plugin unloads)

Janmm14 replied:
resolved?

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java:
```diff
@@ -217,6 +245,9 @@ public Difficulty difficulty() {
     @Override
     public Collection<EntityPlayer> onlinePlayers() {
+        if (!mainThread()) {
+            this.logger.warn("Getting players not from world's thread. This is not safe", new Exception());
```
This does spam GoMint.onlinePlayers()? Should the API be changed?

Janmm14 replied:
The players map is not thread safe and the method only returns an unmodifiable view and does not copy.

Not sure which solution to this is best. (concurrent map? defensive copy?)

geNAZt replied:
Problem is that only the view itself is unmodifiable. This means that when `this.players` changes it still will CME whilst onlinePlayers() iterates. I guess a defensive copy would be best here since it would also get rid of the problem when players are in two worlds (which they could when they teleport whilst querying for GoMint#onlinePlayers())

Janmm14 replied:
Another possible option would be to maintain another serverwide set of players just for this api.
Defensive copy could also CME.

geNAZt replied:
another idea would be to use a stream or consumer and change that API all together

Janmm14 replied:
> another idea would be to use a stream or consumer and change that API all together

You mean that it schedules the player consumer on every world thread sync?

geNAZt replied:
yes

Janmm14 replied:
What about making `GoMint.onlinePlayers()` just return `this.playersByUUID.values()`?

I wil still add the player consumer on every world thread, that sounds potentially useful for plugins to me.

Edit: I justdid this. Decide if you're ok with it or not.

geNAZt replied:
Thats ok, could we reduce the API a bit by making the Consumer based method a overload from the `onlinePlayers()`? So we end up with `onlinePlayers(Consumer) GoMint` and `onlinePlayers() Collection`

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldManager.java:
```diff
@@ -161,13 +140,35 @@ private World loadLevelDBWorld(File path) throws WorldLoadException {
+            synchronized (unloadListener) {
+                if (!unloadListener.done) {
+                    try {
+                        unloadListener.wait();
```
Please use a CountDownLatch here

Janmm14 replied:
Thanks

Janmm14 replied:
resolved?

Commit:
Merge branch 'master' into world-multithreading

Janmm14 committed:
Move world threads to ExecutorService.

Janmm14 committed:
Use CountDownLatch in WorldManager#close, add unloadInternal method to WorldAdapter to actually wait until whole world unloading is done.

Janmm14 committed:
Remove unneeded world null check in player teleportation

Janmm14 committed:
Add world null check for PlayerConnection#addToSendQueue

Janmm14 committed:
Fix tests once again by waiting until the world thread started and initialized threadId (otherwise world.mainThread returns true and during the tests the world thread starts and we get exceptions).

Janmm14 committed:
Don't warn about internal event listener classes, fix PluginWorldConfigManager event listener methods being private

Janmm14 committed:
Introduce private method isListenerMethod in EventManager, don't register private and protected methods.

Janmm14 committed:
Fix javadoc in event api and impl stating ignoreCancelled behaving the exact other way round than implementation handles it currently, current impl behaviour (true = ignore events if they are cancelled) is equal to bukkit, we should not invert that
Fix typo in EventHandlerList
Add void return type check to EventManager listener method detection (require void return type, as EventCallerClassCreator assumes already)
Improve documentation of EventHandler and EventPriority

Janmm14 committed:
Add registerListener method with Predicate argument
Rename registerListener with world string set argument to registerActiveWorldsListener and remove worlds set argument and use new method Plugin#eventInActiveWorlds
Refactor activeWorlds to activeWorldsSnapshot
Make many methods final in Plugin class
Set worlds.yml isBlockList to true by default, warn if its false and worldList is empty.

Commit:
Merge branch 'master' into world-multithreading

Janmm14 committed:
Improve WorldTestUtil code by replacing waitObject and done boolean with a CountDownLatch(1)

Janmm14 committed:
Add World#iterateAllEntities method (to get around raw type warnings with other method)

Janmm14 committed:
Force synchronized world setting
world synchronized disconnect on login event cancel
Check entity for null as well in PlayerConnection#addToSendQueue

Janmm14 committed:
Improve iterateAllEntities method, also add it to Chunk
Sync player cleanup entity iterating with worlds.

Janmm14 committed:
Add UnsafeWorldAsyncAccessWarning exception, unify async access warnings, stack trace already shows the attempted action
Add async access warnings to Chunk

Janmm14 committed:
Little fixes to chunk javadoc

Janmm14 committed:
Add GoMint#onlinePlayersIterate(Consumer<EntityPlayer>) method, scheduling into each world's thread
GoMint#findPlayerByName now uses playersByUUID.values() instead of each world's onlinePlayers
GoMint#onlinePlayers now uses playersByUUID.values()

Janmm14 committed:
Start thread safety in command execute system.

add EntityPlayer#dispatchCommand with optional world scheduling and outputConsumer argument
console commands are forced to run on gomint main thread, player commands are forced to run on world thread
add CommandSender#world()
remove generics from PlayerCommandSender, fix it on EntityPlayer
add UnsafeWorldAsyncAccessException when we don't want to warn but abort on async access

Janmm14 committed:
Add two async world access warnings to Entity, fix javadoc order of impl class Values

Janmm14 committed:
implementation of GoMint returns implementation types where possible, WorldManager returns WorldAdapter instead of World

Janmm14 committed:
Command API rework with support for async command execution. Removes sync CommandOutput result methods.
Commands are by default scheduled on player's world thread (gomint main thread for console commands). This means that usually the thread is already correctly chosen, as most comands affect the world the player is in currently.
CommandOutput was refactored and is now provided in Command#execute for the execute method to add its messages. When command execution is done, it is required that it calls CommandOutput#markFinished to trigger the Callback (CommandOutput consumer).

Janmm14 committed:
Update provided commands to new command API.

geNAZt commented on gomint-api/src/main/java/io/gomint/command/Command.java:
```diff
@@ -102,14 +105,17 @@ public Command( String name ) {
+     * @param output        the command output to add messages to
      */
-    public abstract CommandOutput execute(CommandSender<?> commandSender, String alias, Map<String, Object> arguments );
+    public abstract void execute(CommandSender<?> commandSender, String alias, Map<String, Object> arguments, CommandOutput output);
```
Why is it needed to move the output to a preallocated argument?

Janmm14 replied:
Constraint due to https://github.com/gomint/gomint/pull/727#discussion_r571708144 and to supply the CommandOutput consumer to the CommandOutput.

geNAZt commented on gomint-api/src/main/java/io/gomint/command/CommandOutput.java:
```diff
@@ -11,122 +11,81 @@
+ * process it in the proper manner. Sending chat messages for the command result is NOT recommended and should NEVER be
+ * done.
+ * <br><br>
+ * <b>You need to call {@linkplain #markFinished()} when your command has finished.</b>
```
Not a big friend of markFinished() what happens when i don't call it? Why is it needed?

Janmm14 replied:
To allow finishing command execution in a different thread. It is just a draft.

If you don't call it, Consumer<CommandOutput> will not be called and therefore no command output is sent to the player.

Janmm14 replied:
I have an idea to change it so it does not need to be called always.

Janmm14 replied:
Are you satisfied with the command api changes in 4df9843?

geNAZt replied:
yes

geNAZt commented on gomint-server/src/main/java/io/gomint/server/command/CommandManager.java:
```diff
@@ -129,29 +147,21 @@ public void executeSystem(String line) {
-     * @param command        which should be executed
-     * @param outputConsumer the command output consumer
+     * @param command        which should be executed (includes leading {@code /})
+     * @param outputConsumer the command output consumer, no guarantee is given in which thread it is executed
```
The command and outputConsumer should be executed in the entities world thread (or in the gomint mainthread for system commands) this should be ensured with a Precondition safeguard

Janmm14 replied:
The command itself will be executed in entities world thread. Due to my idea of async command execution end, the outputCosumer is called when CommandOutput#markFinished() is called, which might happen in a different thread, depending on the command's implementation.

Janmm14 committed:
Fix wrong scheduling of teleportation

Janmm14 committed:
Don't require CommandOutput#markFinished() call for sync only commands

Calling CommandOutput#markFinished() is now only required when calling markAsync(), which is required to be called if plugins want to add to the output asynchroniously.
If CommandOutput is not marked as async, the CommandManager will call markFinished().

Calling markFinished without calling markAsync is not problematic, but not needed.

Janmm14 committed:
Rename onlinePlayersIterate to onlinePlayers

Janmm14 committed:
Move world thread check to Precondition.safeWorldAccess(World)

Janmm14 committed:
Fix double player ticking in world ticking, replace more world main thread checks with Precondition.safeWorldAccess

Janmm14 committed:
Thread safe permission and permission-group system, add helper method for String.endsWith(char)

Janmm14 committed:
Call PlayerWorldLeaveEvent on disconnect, add from/to world to PlayerWorldJoinEvent and PlayerWorldLeaveEvent.

Janmm14 committed:
Force sync scoreboard set/remove to player

Janmm14 committed:
Revert "Thread safe permission and permission-group system, add helper method for String.endsWith(char)"

This reverts commit ad24c89f

Janmm14 committed:
Synchronized permission and permission-group access, add helper method for String.endsWith(char)

Janmm14 committed:
Synchronize scoreboard's access/update, Make Scoreboard.ScoreboardLine a static inner class

Janmm14 committed:
Allow async packet send queue adding by making access to it synchronized

Janmm14 committed:
Remove synchronize check in PacketCommandRequestHandler as addToSendQueue is now thread safe

Janmm14 committed:
Merge branch 'master' into world-multithreading

Commit:
Merge branch 'master' into world-multithreading

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java:
```diff
@@ -663,9 +663,6 @@ private void tickLoop() {
     @Override
     public void update(long currentTimeMS, float dT) {
-        for (io.gomint.server.entity.EntityPlayer player : this.players.keySet()) {
-            player.update(currentTimeMS, dT);
```
Who does update the players now?

Janmm14 replied:
The entity manager does this already, just like now.

Janmm14 committed:
Add activeWorldsPlayers API

Janmm14 committed:
Merge remote-tracking branch 'origin/world-multithreading' into world-multithreading

Janmm14 committed:
Commands are restricted to plugin's active worlds by default. Allow for global commands, also add option to disallow console.

Janmm14 committed:
Fluent getters for Command, SubCommand now checks active worlds also for command map sending

Janmm14 committed:
Add UncaughtExceptionHandler to threads.

Janmm14 committed:
Let environment variable LOG_LEVEL define log level

Janmm14 committed:
Many Fixes and a couple optimizations
- Move server.executorService() to server.asyncScheduler() to be notified about exceptions happening
- move player to world tick earlier, add newConnections list to each world for that
- fix npe in login as spawnLocation can be null
- network manager untick queue is now just an arraylist (enough)
- Change closing behaviour:
  - close network later to be able to kick players
  - disallow new connections when shutting down
  - players get kicked with "Server closed" instead of just connection closing.
 - fix potential watchdog npe when thread is dead
 - report exceptions in world tick and shutdown server
 - uncaught exception handler reports thread id as well
 - added a couple more debug log messages

Janmm14 committed:
Revert @version javadoc tag changes

Janmm14 committed:
Remove unneeded markFinished from TestCommand

Janmm14 committed:
ignore files of running gomint in release folder

Janmm14 committed:
@author Janmm14

Janmm14 committed:
Add toString with super class accessors to events. Fixes #719

geNAZt commented on gomint-api/src/main/java/io/gomint/GoMint.java:
```diff
@@ -249,15 +287,10 @@ static GoMint instance() {
-     * @return tickrate of this server
+     * @return GoMint for chaining
      */
-    double tps();
```
The TPS getter should be present. There are plugin events on that ticker. Maybe extends the documentation that this TPS only affects join and connect events.

Janmm14 replied:
readded

geNAZt commented on gomint-api/src/main/java/io/gomint/command/CommandOutput.java:
```diff
@@ -164,4 +123,49 @@ public boolean success() {
+     * @see #markAsync() 
+     * @see #isFinished() 
+     */
+    public synchronized void markFinished() {
```
This should be chainable

Janmm14 replied:
No, it should not be. Anything happening after markFinished call is essentially ignored, as internally markFinished sends the results.

geNAZt commented on gomint-api/src/main/java/io/gomint/event/entity/projectile/ProjectileHitBlocksEvent.java:
```diff
@@ -49,7 +49,7 @@ public ProjectileHitBlocksEvent(Set<Block> blocks, EntityProjectile<?> projectil
      * @deprecated Use {@link #entity()} instead
      */
-    @Deprecated(since = "2.0", forRemoval = true)
+    @Deprecated(forRemoval = true)
```
Can we remove this now? Since we have API breakage in so many places this should also go now

Janmm14 replied:
Ok

Janmm14 replied:
removed

geNAZt commented on gomint-api/src/main/java/io/gomint/plugin/Plugin.java:
```diff
@@ -115,28 +133,68 @@ public void onUninstall() {
+     * @see #registerListener(EventListener, Predicate)
+     * @since 2.0
+     */
+    public final Plugin registerActiveWorldsListener(EventListener listener) {
```
Could we only use one register? Is there a reason why the default registerListener is not per default a active Worlds listener?

Janmm14 replied:
Good idea

Janmm14 replied:
I think I named this method so, because the registerListener method with the custom Predicate argument does add its own active worlds check. Should the method for listener registration without predicate be renamed to registerGlobalListener or removed?

geNAZt replied:
Can you use the same name but without the predicate argument?

Janmm14 replied:
We currently have `registerListener(EventListener)`, `registerListener(EventListener, Predicate)` and `registerActiveWorldsListener(EventListener)`.
Can you explain what you want renamed exactly please, I'm a little confused.

geNAZt replied:
registerActiveWorldsListener(EventListener) -> registerListener(EventListener)

which then calls 

registerListener(EventListener, Predicate) with null which internally always has the world checking predicate and the one given from the plugin

geNAZt commented on gomint-api/src/main/java/io/gomint/plugin/Plugin.java:
```diff
@@ -179,31 +237,105 @@ public final InputStream resourceAsStream(String name ) {
+     */
+    @Nullable
+    public final Set<String> activeWorldsSnapshot() {
+        return this.activeLoadedWorlds == null ? null : ImmutableSet.copyOf(this.activeLoadedWorlds);
```
This does not check for loaded worlds. Either the comment is wrong or the check is missing

Janmm14 replied:
The activeLoadedWorlds set is updated by gomint when worlds are (un-)loaded:
https://github.com/gomint/gomint/blob/cbcfd86358a988308d0c869a86e615acc2d7a0a9/gomint-server/src/main/java/io/gomint/server/plugin/PluginWorldConfigManager.java

geNAZt commented on gomint-api/src/main/java/io/gomint/plugin/PluginManager.java:
```diff
@@ -75,9 +78,45 @@
+     * @see #registerListener(Plugin, EventListener, Predicate)
+     * @since 2.0
+     */
+    PluginManager registerActiveWorldsListener(Plugin plugin, EventListener listener);
```
Same as above (why do we seperate the listeners)

geNAZt commented on gomint-api/src/main/java/io/gomint/world/World.java:
```diff
@@ -314,4 +325,30 @@
+     *
+     * @return target tickrate of this world
+     */
+    int targetTPS();
```
Should there be a setter so we can have dynamic ticking worlds (like having one tick faster than another one)?

Janmm14 replied:
Possible to add this, but I won't add it to this PR.

geNAZt commented on gomint-server/src/main/java/io/gomint/server/command/CommandManager.java:
```diff
@@ -41,15 +44,36 @@
+
+    static {
+        try {
+            COMMAND_OUTPUT_CONSTRUCTOR = CommandOutput.class.getDeclaredConstructor(Consumer.class);
```
Why is this a reflection and not a direct invocation of the constructor with new?

Janmm14 replied:
Because new CommandOutput should not be able to be created by plugins anymore and its in a different package.

geNAZt replied:
Well then wouldn't it make sense to pull CommandOutput into gomint-server and expose only a interface like blocks and items do?

Janmm14 replied:
oh yeah

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/command/CommandManager.java:
```diff
@@ -103,30 +127,52 @@ public CommandManager() {
+            if (world.mainThread()) {
+                dispatchCommand0(sender, command, output);
+            } else {
+                world.syncScheduler().execute(() -> {
```
This should error, there should be no clear way to delay command execution (this could lead to the player not being connected / on that world anymore)

Janmm14 replied:
changed

geNAZt commented on gomint-api/src/main/java/io/gomint/command/annotation/Scope.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * @see io.gomint.command.Command#activeWorldsOnly(boolean)
+     */
+    boolean activeWorldsOnly() default true;
```
Why can plugins disable this? Shouldn't this be always the case?

Janmm14 replied:
Just as plugins can still register listeners with their own predicate or with no predicate at all, it is possible to not restrict commands to specific worlds.

geNAZt replied:
Thats where we collide i guess. This world restriction stuff should not be system where the plugin developer has a say. It should be a tool for operators which can limit every plugin (if it explicitly supports it or not) to specific worlds.

Janmm14 replied:
Ok, then I will change it to not giving plugin devs a say.

geNAZt commented on gomint-server/src/main/java/io/gomint/server/command/cli/MemoryDumpCommand.java:
```diff
@@ -40,8 +40,7 @@ public CommandOutput execute(CommandSender<?> commandSender, String alias, Map<S
         }
-
-        return null;
+        output.markFinished();
```
This is not async thus doesn't need markFinished

Janmm14 replied:
Yes, will change. Luckily invoking it does not break anything when its the last thing a command does with the output.

Janmm14 replied:
removed

geNAZt commented on gomint-server/src/main/java/io/gomint/server/network/NetworkManager.java:
```diff
@@ -424,23 +463,14 @@ public int port() {
-    public void shutdown() {
-        LOGGER.info("Shutting down networking");
-        if (this.socket != null) {
-            this.socket.close();
```
Who closes the socket now?

Janmm14 replied:
This was actually closing the socket too early on shutdown, not allowing kicks to be sent.
Socket is closed later in NetworkManager's close method, which is invoked later in the overall shutdown process.
https://github.com/gomint/gomint/pull/727/files/cbcfd86358a988308d0c869a86e615acc2d7a0a9#diff-14f9cdd18c69f2343186ba0b1124abcbca4975f06b54fe57f2e48fd50d311343R269

geNAZt commented on gomint-api/src/main/java/io/gomint/scheduler/WorldScheduler.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @return the world synchronized tasks will be associated to
+     */
+    World world();
```
This can be null, please document

Janmm14 replied:
When can this be null?
When plugins for whatever reason save WorldScheduler instances forever and access it after the world was unloaded?
Then the whole WorldScheduler instance gets invalid and useless.

geNAZt replied:
Yes thats the case. We should document that this will become null when the world unloads and thus storing a instance of this scheduler is NOT recommended

Janmm14 replied:
documented

geNAZt commented on gomint-server/src/main/java/io/gomint/server/scheduler/PluginWorldSchedulerAdapter.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public Task execute(Runnable runnable) {
+        return this.pluginSchedulerAdapter.execute(world(), runnable);
```
Please check for null from world(). This will NPE in the adapter

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/scheduler/PluginWorldSchedulerAdapter.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public Task schedule(Runnable runnable, long delay, TimeUnit timeUnit) {
+        return this.pluginSchedulerAdapter.schedule(world(), runnable, delay, timeUnit);
```
Please check for null from world(). This will NPE in the adapter

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/scheduler/PluginWorldSchedulerAdapter.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public Task schedule(Runnable runnable, long delay, long period, TimeUnit timeUnit) {
+        return this.pluginSchedulerAdapter.schedule(world(), runnable, delay, period, timeUnit);
```
Please check for null from world(). This will NPE in the adapter

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/util/Precondition.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @param world the world to enforce sync access to
+     */
+    public static void safeWorldAccessHard(@Nullable WorldAdapter world) {
```
Can we rename this? This is only used to allow the I/O thread. This could be a parameter "allowForIOThread" in safeWorldAccess and the mainThread() check

Janmm14 replied:
will do

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldManager.java:
```diff
@@ -7,34 +7,57 @@
     private final GoMintServer server;
-    private Object2ObjectMap<String, WorldAdapter> loadedWorlds;
+    private Map<String, WorldAdapter> loadedWorlds;
+    private ThreadGroup threadGroup = new ThreadGroup(THREAD_GROUP_NAME);
```
Those should be final

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldManager.java:
```diff
@@ -7,34 +7,57 @@
-    private Object2ObjectMap<String, WorldAdapter> loadedWorlds;
+    private Map<String, WorldAdapter> loadedWorlds;
+    private ThreadGroup threadGroup = new ThreadGroup(THREAD_GROUP_NAME);
+    private ExecutorService executorService = Executors.newCachedThreadPool(new ThreadFactory() {
```
Those should be final

Janmm14 replied:
done

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java:
```diff
@@ -476,18 +517,189 @@ private void initGamerules() {
+                    .includeWorlds()
+                    .upload();
+                this.running.set(false);
+                this.server.shutdown();
```
Should one world crashing take down the whole instance? There might be other worlds being fine? Maybe unloading the world as a first mitigation would be better

Janmm14 replied:
Possible, I just decided to readd that behaviour as thats how it currently works.

Janmm14 replied:
Behaviour changed

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java:
```diff
@@ -1669,7 +1900,55 @@ public int timeAsTicks() {
+            return true;
+        } else {
+            long id = Thread.currentThread().getId();
+            return this.threadId == id || this.asyncAllowedThreadId == id;
```
Please merge the soft and hard method into one with a parameter

Janmm14 replied:
done

Janmm14 committed:
Address some requested changes, add isRunning to server & worlds api

Janmm14 committed:
Remove unneeded markFinished call

geNAZt commented on gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java:
```diff
@@ -1916,23 +1915,25 @@ public WorldScheduler scheduler() {
         } else {
             long id = Thread.currentThread().getId();
-            return this.threadId == id;
+            return this.threadId == id || allowForIOThread && this.asyncAllowedThreadId == id;
```
allowForIOThread && this.asyncAllowedThreadId == id need to be in ( ) ?

Janmm14 replied:
&& binds stronger than ||

Janmm14 committed:
Address more requested changes.

- Remove Command scope system again, Commands are now always bound to active worlds.
- Event listeners are now always bound to active worlds.
- Removed some more unneeded calls to CommandOutput#markFinished

Janmm14 committed:
Update javadoc in CommandManager to reflect current behaviour of dispatchCommand

Janmm14 committed:
Don't shutdown server on world crash. Kick remaining players on world stop.

Janmm14 committed:
Code style: Remove an unneeded cast
